target/linux/octeon/base-files/lib/upgrade/platform.sh

```diff
@@ -1,116 +1,110 @@
+		cp -af "$UPGRADE_BACKUP" "/mnt/$BACKUP_FILE"
+		umount "/mnt"
+	}
 }
 
 platform_copy_config() {
 	case "$(board_name)" in
-	erlite)
-		mount -t vfat /dev/sda1 /mnt
-		cp -af "$UPGRADE_BACKUP" "/mnt/$BACKUP_FILE"
-		umount /mnt
-		;;
-	itus,shield-router)
-		mount -t vfat /dev/mmcblk1p1 /mnt
-		cp -af "$UPGRADE_BACKUP" "/mnt/$BACKUP_FILE"
-		umount /mnt
-		;;
+		erlite)
+			platform_octeon_copy_config "/dev/sda1"
+			;;
+		itus,shield-router)
+			platform_octeon_copy_config "/dev/mmcblk1p1"
+			;;
 	esac
 }
 
 platform_do_flash() {
-	local tar_file=$1
-	local board=$2
-	local kernel=$3
-	local rootfs=$4
-
-	local board_dir=$(tar tf "$tar_file" | grep -m 1 '^sysupgrade-.*/$')
-	board_dir=${board_dir%/}
-	[ -n "$board_dir" ] || return 1
-
-	mkdir -p /boot
-
-	if [ $board = "itus,shield-router" ]; then
-		# mmcblk1p1 (fat) contains all ELF-bin images for the Shield
-		mount /dev/mmcblk1p1 /boot
-
-		echo "flashing Itus Kernel to /boot/$kernel (/dev/mmblk1p1)"
-		tar -Oxf $tar_file "$board_dir/kernel" > /boot/$kernel
-	else
-		mount -t vfat /dev/$kernel /boot
-
-		[ -f /boot/vmlinux.64 -a ! -L /boot/vmlinux.64 ] && {
-			mv /boot/vmlinux.64 /boot/vmlinux.64.previous
-			mv /boot/vmlinux.64.md5 /boot/vmlinux.64.md5.previous
-		}
+	local sysupgrade_file="$1"
+	local kernel_file="$2"
+	local kernel_checksum_file="$kernel_file.md5"
+	local boot_device="$3"
+	local root_device="$4"
+	# find a sysupgrade directory by looking inside sysupgrade file
+	local sysupgrade_directory="$(tar tf $sysupgrade_file | awk -F '/' '/^sysupgrade/ {print $1; exit}')"
+	[ -n "$sysupgrade_directory" ] || return 1
+
+	if [ ! -f "$sysupgrade_file" ] && \
+		[ ! -b "$boot_device" ] && \
+		[ ! -b "$root_device" ] ; then
+		return 1
+	fi
 
-		echo "flashing kernel to /dev/$kernel"
-		tar xf $tar_file $board_dir/kernel -O > /boot/vmlinux.64
-		md5sum /boot/vmlinux.64 | cut -f1 -d " " > /boot/vmlinux.64.md5
+	local boot_directory="/boot"
+	mkdir -p "$boot_directory"
+	mount -t "vfat" "$boot_device" "$boot_directory"
+	# backups current kernel file to a kernel file with .previous postfix
+	if [ -f "$boot_directory/$kernel_file" ] && \
+		[ ! -L "$boot_directory/$kernel_file" ] && \
+		[ -f "$boot_directory/$kernel_checksum_file" ] && \
+		[ ! -L "$boot_directory/$kernel_checksum_file" ] ; then
+		mv "$kernel_file" "$kernel_file.previous"
+		mv "$kernel_checksum_file" "$kernel_checksum_file.previous"
 	fi
 
-	echo "flashing rootfs to ${rootfs}"
-	tar xf $tar_file $board_dir/root -O | dd of="${rootfs}" bs=4096
+	echo "copying kernel to $boot_directory/$kernel_file"
+	tar xf "$sysupgrade_file" "$sysupgrade_directory/kernel" -O > "$kernel_file"
+	md5sum "$kernel_file" | cut -f1 -d " " > "$kernel_checksum_file"
+
+	echo "flashing rootfs to $root_device"
+	tar xf "$sysupgrade_file" "$sysupgrade_directory/root" -O | dd of="$root_device" bs=4096
 
 	sync
-	umount /boot
+	umount "$boot_directory"
+	return 0
 }
 
 platform_do_upgrade() {
-	local tar_file="$1"
-	local board=$(board_name)
```

Comment:
still using this variable in this function...

```
not yet supported on $board.
```

definitely an improvement from the originals once you iron out the finer quirks... the 'if/else' for board detect was particularly problematic...

- not sold on awk / replacement of board_dir for consistency sake...
- lack of md5 on itus will skip the first kernel.previous?
- triple check your quotings, variable names

but all in all... a step in the right direction I think...

Reply:
> not sold on awk / replacement of board_dir for consistency sake...

well, it is also to reduce logic.
current:
1. we check tar file
2. grep for matches of sysupgrade directory name
3. do subtitution to get a correctly formatted sysupgrade directory name

proposed:
1. we check tar file
2. use single awk to get sysupgrade directory name

> lack of md5 on itus will skip the first kernel.previous? ( benign? )
i added one more extra step for creating md5 checksum for situations where it does not exist.
maybe we should add a failure in case md5 checksum does not match what we have in file?

> triple check your quotings, variable names
seem to be fine now. checked by shellcheck and fixed its complains about SC2155.

please check it again

thank you

Reply:
thanks, i'm a novice around these parts... so i'll have to let the pro's take it from here...

Reply:
each PR is self descriptive and contain certain change that is not directly related to other changes except that they're in the same tree.

and smaller changes are and easier to understand - easier to get accepted and get merged.